### ISSUE TITLE
Codes for self-attention version

### ISSUE THREAD

**Issue by yaokl-nju** [wokas36/DFNets#1]
Hi,

  I'm very interesting in your work, could you provide the codes for self-attention version?

Thank you very much!

**Comment by wokas36** [comment]
Hi klyao,

Thanks for interesting our work, could you please find the DFNets with self-attention version (dfnet-att) using the following link [DFNet-ATT](https://github.com/wokas36/DFNets/tree/master/dfnet-att).

**Comment by yaokl-nju** [comment]
Hi wokas36,

Thank you very much for releasing the codes! I have some questions for reproducing results, and it's about how to select final testing results:
1. Select the testing results at the final epoch (200)? If so, I run both dfnet and dfnet-att with 100 runs, and the results are not consistent with those in the paper by a little large gap. I don't know what results into the differences.
2. If we select the testing results according to the validation loss or accuracy, the results would be worse than option 1.

Can you provide a complete version for reproducing results in the paper? That will help a lot!

I will be very appreciate if you can answer my questions.

**Comment by wokas36** [comment]
Hi klyao,

Extremely sorry for the late reply (I was traveling in last two weeks and limited access of internet). Thanks in advance for your valuable suggestions and feedback for improving our code quality and results. Testing results not always belong with the final epoch, it’s usually attached with the epoch where the model achieved the best accuracy in both training and testing. In addition, we can stop the training when the accuracy is not improving by simply adding EarlyStopping OR saving model parameters (weights) at the optimal stage by using ModelCheckpoint. If u need to reproduce our results, please change the hyperparameter setting of `mu` to 70 for dfnet and 100 for dfnet-att respectively for successfully reproducing the similar results. However, there may have tiny differences due to the stochasticity of the models (due to randomly initializing the weights and biases). If you need to reproduce the results with training/validation/testing splits, you need to reconfigure the hyperparameter settings in advance w.r.t validation dataset. But it may take some times, because our hyperparameter space is larger than the usual settings (we introduce 7 to 8 new hyperparameter  settings for dfnets). You can use [dfnet](https://github.com/wokas36/DFNets/tree/master/dfnet) for reproducing the results with training/validation/testing splits. Hope that our reply can help you.
